protocol/auditlog.go

```diff
@@ -0,0 +1,180 @@
+// for "catching up" on a history in case an auditor misses epochs
+func (l *ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
+	oldSTRs map[uint64]*DirSTR, latestSTR *DirSTR) error {
+
+	// error if we want to create a new entry for an addr we already know
+	if l.IsKnownDirectory(addr) {
+		return ErrAuditLog
+	}
+
+	// create the new directory history
+	h := newDirectoryHistory(signKey, latestSTR)
+
+	startEp := uint64(0)
+	endEp := latestSTR.Epoch
+
+	// add each old STR into the history
+	for ep := startEp; ep < endEp; ep++ {
+		str := oldSTRs[ep]
+		if str == nil {
+			return ErrMalformedDirectoryMessage
+		}
+		h.snapshots[ep] = str
+	}
+
+	l.histories[addr] = h
+
+	// FIXME: verify the consistency of each new STR
+	return nil
+}
+
+// Update verifies the consistency of a newly observed STR newSTR for
+// the directory addr, and inserts the newSTR into addr's directory history
+// if the checks (i.e. STR signature and hash chain verifications) pass.
+// Update() returns nil if the checks pass, and the appropriate consistency
+// check error otherwise. Update() assumes that Insert() has been called for
+// addr prior to its first call and thereby expects that an entry for addr
+// exists in the audit log l.
+// FIXME: pass Response message as param
+func (l *ConiksAuditLog) Update(addr string, newSTR *DirSTR) error {
+
+	// error if we want to update the entry for an addr we don't know
+	if !l.IsKnownDirectory(addr) {
+		return ErrAuditLog
+	}
+
+	h := l.histories[addr]
+
+	if err := verifySTRConsistency(h.signKey, h.latestSTR, newSTR); err != nil {
+		return err
+	}
+
+	// update the latest STR
+	h.snapshots[h.latestSTR.Epoch] = h.latestSTR
+	h.latestSTR = newSTR
+	return nil
+}
+
+// GetObservedSTRs gets the observed STR for the CONIKS directory
+// address for a directory history entry indicated in the
+// AuditingRequest req received from a CONIKS client,
+// and returns a tuple of the form (response, error).
+// The response (which also includes the error code) is supposed to
+// be sent back to the client. The returned error is used by the auditor
+// for logging purposes.
+//
+// A request without a directory address or with an epoch greater than
+// the latest observed epoch of this directory is considered malformed,
+// and causes GetObservedSTRs() to return a
+// message.NewErrorResponse(ErrMalformedClientMessage) tuple.
+// GetObservedSTRs() returns a message.NewSTRHistoryRange(strs) tuple.
+// strs is a list of STRs for the epoch range [ep,
+// l.histories[req.DirectoryAddr].latestSTR.Epoch], where ep is the epoch for
+// which the client has requested the observed STR; i.e. if ep == the latest epoch,
+// the list returned is of length 1.
+// If the auditor doesn't have any history entries for the requested CONIKS
+// directory, GetObservedSTRs() returns a
+// message.NewErrorResponse(ReqUnknownDirectory) tuple.
+func (l *ConiksAuditLog) GetObservedSTRs(req *AuditingRequest) (*Response,
```

Comment:
> I think it could also happen if the EpochDeadline is very small, say, 5s, and the client has a high network latency, even if it does monitoring at the beginning of en epoch. But I agree that we shouldn't make things more complicated. What you said makes sense to me.

This is a really good point, though. So, maybe what makes sense is to change the `AuditingRequest` message to have a `startEpoch` and `endEpoch` instead, and if the client wants one STR, it sets both params to the same epoch? And if it needs to catch up, it can also indicate this with this message format.

> Do you mean the client fetches a range of STR from the server, verifies them, and compare the newest STR in the range with the auditors?

Right, although this isn't the full picture. The full flow is: (1) the client fetches an STR range from the server. (2) the client does the hash chain check, and if it passes the check, the client accepts and caches the latest STR from the server. (3) the client requests the same STR range from an auditor. (4) the client verifies that range as in step 2. (5) if the client independently accepts the latest STRs from both, it then compares only the latest STR. If there has been a fork at any point in time, the STRs *will* differ because somewhere along the way, the `prevHash` will be different. So there's no need for the client to compare every STR in the range. It's also important to note that this flow will only happen if the client misses epochs, so it should be a less common operation.